### ISSUE TITLE
Using taghelpers in razor views using a model type which is loaded runtime

### ISSUE THREAD

**Issue by DavidEmanuelsen** [aspnet/Mvc#4497]
Hi,

I have created an application which uses a custom fileprovider for loading razor views from disk / database, and these views uses models coming from an assembly which is loaded runtime. I've used the RazorViewEngineOptions.CompilationCallback to load this assembly and the compilation of the views works - except for when using `asp-for` etc. (The compilation callback was implemented as per the example given in #4423)

The model object is created using Activator.CreateInstance where the type is coming from the external assembly.

The errors I'm getting when using taghelpers are:
- `Type expected`
- `Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type`

These errors don't contain much detail, but I'm assuming the problem is related to the model type which is loaded runtime. Any suggestions on how I can get this to work is greatly appreciated.

**Comment by dougbu** [comment]
If any of the above errors are actually thrown exceptions, please provide the stack traces.

**Comment by Eilon** [comment]
@DavidEmanuelsen indeed, where are you seeing these errors, and can you show us any relevant code from your app?

**Comment by DavidEmanuelsen** [comment]
I have project containing the model to be used which implements an interface which is also know by the running application. This project / class library is build seperately, and in the main application I create an instance of the model like this:

```
Assembly externalAssembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(assemblyPath);
var type = Type.GetType(serviceNamespace 
            + "." + serviceImplementationName 
            + "," + serviceNamespace);

return (IServiceModel)Activator.CreateInstance(type);
```

The compilation callback looks like this:
```
services.Configure<RazorViewEngineOptions>(options =>
{
        var previous = options.CompilationCallback;
        options.CompilationCallback = (context) =>
        {
               previous?.Invoke(context);
               MetadataReference[] references = serviceAssemblyReferences.ToArray();
                    
               context.Compilation = context.Compilation.AddReferences(references);
         };

         options.FileProviders.Add(new CustomFileProvider());
});
```

Where the serviceAssemblyReferences-list is populated with new items as needed before returning from the controller (creating new items using `MetadataReference.CreateFromFile(path)`)

I'm having some issues debugging the application at the moment (not able to run it directly from VS, and when I use dotnet publish I can't attach to the running process), but I will update with stacktraces as soon as I can get them. Currently I'm only getting what the developer exception page is showing me when using `<input asp-for="FirstName" type="text" />` or `@Html.EditorFor(m => m.FirstName)`

![image](https://cloud.githubusercontent.com/assets/18324505/14631706/0298ad96-0615-11e6-8ac1-32a2114bf88d.png)

**Comment by DavidEmanuelsen** [comment]
Just realized I can give the type using `@Html.EditorFor<MyModelType, int>(...)`.. How can the same be done for the taghelpers?

**Comment by Eilon** [comment]
@DavidEmanuelsen - any chance you can replicate this scenario without the custom assembly loads, file system, etc.? In other words, can you create a regular CSHTML file on disk that uses regular models in regular CS files? I'm guessing you'll hit the exact same error, but it might be easier to replicate and debug from that.

**Comment by dougbu** [comment]
> Just realized I can give the type using `@Html.EditorFor<MyModelType, int>(...)`. How can the same be done for the tag helpers?

What is the advantage you see in calling `EditorFor()` with explicit type parameters? Suggest focusing on why can't the compiler infer those parameters from the expression.

In any case you'll need parentheses around the generic method call: `@(Html.EditorFor<MyModelType, int>(...))`.

Tag helpers have no equivalent to this. While it's possible to explicitly create a `ModelExpression` containing the correct information e.g. `CreateModelExpression(m => m.Name)` that instance can't be assigned to an `asp-for` attribute value -- that value is always _part_ of a lambda expression in the generated code.

**Comment by DavidEmanuelsen** [comment]
I've tried reproducing it without any assembly loads etc, but no luck. I'm only able to reproduce it when loading the assembly containing the model runtime. I've created a simple project which illustrates my issue.

Controller action:
```
public ActionResult Index()
{
     return View(Activator.CreateInstance(Type.GetType("ModelProject.MyModel,ModelProject")));
}
```

View (now contained within the running application instead of loading from a custom location):
```
@model ModelProject.MyModel
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers

<input asp-for="MyProperty" type="text" />
```

Startup:
```
    public class Startup
    {
        public static List<MetadataReference> serviceAssemblyReferences = new List<MetadataReference>();

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc();
            services.Configure<RazorViewEngineOptions>(options =>
            {
                var previous = options.CompilationCallback;
                options.CompilationCallback = (context) =>
                {
                    previous?.Invoke(context);

                    context.Compilation = context.Compilation.AddReferences(serviceAssemblyReferences.ToArray());
                };
            });
        }

        public void Configure(IApplicationBuilder app)
        {
            app.UseDeveloperExceptionPage();
            app.UseStaticFiles();
            app.UseMvcWithDefaultRoute();
        }
        
        public static void Main(string[] args)
        {
            var ctx = new MyAssemblyLoadContext();
            AssemblyLoadContext.InitializeDefaultContext(ctx);

            var host = new WebHostBuilder()
                .UseDefaultHostingConfiguration(args)
                .UseIISIntegration()
                .UseKestrel()
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }

    public class MyAssemblyLoadContext : AssemblyLoadContext
    {
        protected override Assembly Load(AssemblyName assemblyName)
        {
            string path = "C:\\ServiceAssemblies\\" + assemblyName.Name + ".dll";
            Startup.serviceAssemblyReferences.Add(MetadataReference.CreateFromFile(path));

            return LoadFromAssemblyPath(path);
        }
    }
```

The ModelProject-assembly is build seperatly, but it only contains one class as follows:
```
public class MyModel
{
     public int MyProperty = 1337;
 }
```

**Comment by Eilon** [comment]
@dougbu can you try to repro and see what this error is about?

**Comment by dougbu** [comment]
@Eilon :ok:

**Comment by dougbu** [comment]
I can reproduce this compilation error without the custom `AssemblyLoadContext`.

The `CSharpCompilation` change is at least partially doing the Right Thing<sup>:tm:</sup>. Without the added `MetadataReference`, Roslyn errors include `The type or namespace name 'ModelProject' could not be found (are you missing a using directive or an assembly reference?)`.

Hard to tell exactly what's going wrong due to the messed-up line numbers in the errors. But based on the columns and the general class of compilation failures, problems seem focused on `ModelExpressionProvider.CreateModelExpression(ViewData, __model => __model.MyProperty);` in the generated code. `ViewData` is of type `ViewDataDictionary<ModelProject.MyModel>` and the lambda is of type `Expression<Func<ModelProject.MyModel, int>>`. The compiler must be having problems inferring these types.

Will continue looking.

/cc @davidfowl any suggestions on where to go next?

**Comment by sebastienros** [comment]
One of Orchard's contributor had to move the compilation callback call to fix a related issue regarding compiling Views with references to dynamically loaded assemblies:

https://files.gitter.im/OrchardCMS/Orchard2/6xVS/roslyn.png

**Comment by Eilon** [comment]
@dougbu thanks, we'll treat this as a bug for RTM but I guess we still don't know exactly where the bug is 😦

**Comment by dougbu** [comment]
I'll see if moving the `Rewrite()` call after `_compilationCallback` is run (as https://files.gitter.im/OrchardCMS/Orchard2/6xVS/roslyn.png shows) improves the situation here too. Might want to consult w/ a Roslyn expert as well.

But this has to wait a day or so...

**Comment by Eilon** [comment]
Yeah no rush.

**Comment by jtkech** [comment]
So, here in such a scenario i needed to load not only the external assembly but also the assemblies of its dependencies which are not part of the current application project.

This by using the system runtime AssemblyLoadContext (also used by roslyn) not to have e.g model / view data type mismatches ...

Then, to make taghelpers working, i needed to update `DefaultRoslynCompilationService`. Not sure it's a good fix in all scenarios but i needed it to get it working.

 I've moved the call of Rewrite(compilation) where syntax trees are rewritten.

            // don't rewrite before _compilationCallback
            //compilation = Rewrite(compilation);

            var compilationContext = new RoslynCompilationContext(compilation);
            _compilationCallback(compilationContext);
            compilation = compilationContext.Compilation;

            // rewrite after _compilationCallback
            compilation = Rewrite(compilation);

Best.

**Comment by jtkech** [comment]
Sorry, just seen you have seen my screenhost on orchard gitter

**Comment by DavidEmanuelsen** [comment]
I've tested the fix suggested by @jtkech and that solves our problem. We'll probably use that as a temporary fix for now, will update here if we run into any issues with it down the line.

**Comment by Eilon** [comment]
@dougbu FYI please remember to add any necessary coverage for https://github.com/aspnet/Razor/issues/763 as well.

**Comment by dougbu** [comment]
Main investigation here is to test whether the scenario works using infrastructure from #4672 and #4739. At the moment, things seem to work using `AssemblyPart`s and .NET 4.5.2. @pakrym is working on issues w/ this scenario when using .NET Core.

Taking this out of Working while that work's in progress.

**Comment by Eilon** [comment]
I talked to @pakrym and @davidfowl and this work will not happen for 1.0.0 RTM. You'll need to use some workaround in the meantime.

**Comment by eimik** [comment]
That's bad news. We were hoping this (for us) quite important issue would get fixed in RTM.

**Comment by Eilon** [comment]
BTW in the meantime we'll look into adding a smaller feature that should unblock most scenarios. The problem with the "full" solution is that it's a fair bit of work that just can't fit on the schedule right now.

@dougbu will post details as he figures them out.

**Comment by NTaylorMullen** [comment]
Here's the options we have:

1. Point users towards the [DefaultRoslynCompilationContext](https://github.com/aspnet/Mvc/blob/e63f094a5fce193de079c9ec117415b5fdbe39d4/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs) (our default `ICompilationService`) and tell them to override the [GetApplicationReferences](https://github.com/aspnet/Mvc/blob/e63f094a5fce193de079c9ec117415b5fdbe39d4/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs#L156) method. There's different forks to this solution, we could leave the class internal and let it truly be a workaround or we could make it public to make it more first class. With this approach, once `ApplicationParts` light up, nothing changes and we have no extra API to maintain in the future (it's already there). Another slightly less appealing route would be to put the method on the `ICompilationService` interface.
2. Add an `AdditionalCompilationReferences` of type `IList<MetadataReference>` to the `RazorViewEngineOptions` where users can add references used to compile Razor views. This would be an add-only first class approach for users to muck with what assemblies a Razor page should have access to. It'd continue to work once `ApplicationParts` light back up since it'd only be additive.
3. Add a `CompilationReferences` of type `IList<MetadataReference>` that is the source of truth for what assemblies are used to compile Razor views. This would make modifying the assemblies used to compile razor views very easy. Issues would arise when `ApplicationParts` come back online because they're both intended to be the source of truth. Also, an API we'd need to maintain like 2.

/cc @Eilon / all what approach would you like me to take?

**Comment by DavidEmanuelsen** [comment]
For our case option 2 would be the most natural, it feels kind of similar to this usage of `RazorViewEngineOptions.CompilationCallback`:
```
options.CompilationCallback = (context) =>
{
     context.Compilation = context.Compilation.AddReferences(serviceAssemblyReferences);
};
```

**Comment by DavidEmanuelsen** [comment]
On second thought.. wouldn't option 2 and the `CompilationCallback` approach be two things trying to solve the same thing? Would there be a need for the option for adding references in the `CompilationCallback` if option 2 were to be implemented?

**Comment by dougbu** [comment]
> wouldn't option 2 and the `CompilationCallback` approach be two things trying to solve the same problem?

Not really. `CompilationCallback()` is called later, after the the Roslyn syntax tree has been created and updated and just before the assembly is generated. It's actually intended to let your code change that final syntax tree. This intent is why we can't just move the call earlier.

**Comment by eimik** [comment]
Ah, thanks for the info, was wondering about that myself. Option 2 seems like a good solution to me.

**Comment by NTaylorMullen** [comment]
I'll work on getting a fix out for option 2 then! Stay tuned :smile:

**Comment by NTaylorMullen** [comment]
51c285574b02e381a0744828d40c2a81001a5a25

**Comment by DavidEmanuelsen** [comment]
@NTaylorMullen 

I've started the move to 1.0.0 and trying to use the new `AdditionalCompilationReferences` to add the `MetadataReference`s that I need. It works fine as long as I populate it during initial setup in Startup.cs, but if I inject the `RazorViewEngineOptions` somewhere else and update the list in the same request as the `MetadataReference` is needed it doesn't always work. It's kind of odd as it seems it works sometimes, but not all.

It might be asking too much, but is it possible to get an example of how to manage the list of additional references correctly? I'm trying to get my head around the compilation works (and also how to use `AssemblyLoadContext` correctly), but I'm struggling a bit to find documentation / examples on these things.

Just to give some context around what we are trying to achieve:
We are creating a platform where users can create Views and logic which we will load from some database after the application has started as needed. So both the viewmodel and the views themselves are located / loaded runtime and are not known by the application when it starts.

Our current plan for loading the views is using a custom `FileProvider,` and for the assemblies we're currently loading them from a db, placing on disk, loading using `AssemblyLoadContext` (the Default one) and also now creating a `MetadataReference` and add it to `AdditionalCompilationReferences` before returning the view. The model is created using `Activator.CreateInstace(..)` after the assembly has been loaded.

I've created a test project which illustrates the behavior we are seeing [here](https://github.com/DavidEmanuelsen/CompilationReferencesTesting). After starting the application, `/Secondary/Index` is not able to resolve the reference to the external assembly if `/Home/Index` has loaded first. However, if I start the application and go `/Secondary/Index` straight away things works as expected (also for subsequent page loads / navigation). This confuses me, what's going on? :)


I guess you are all currently out celebrating the launch, but any help / input / suggestions would be greatly appreciated.

**Comment by pranavkm** [comment]
> but if I inject the RazorViewEngineOptions somewhere else and update the list in the same request as the MetadataReference is needed it doesn't always work

Modifying any of the option types after Startup has done is not supported. Types are free to cache individual properties of the options, so updating it would not necessarily work all the time.
For your scenario, I'd recommend not relying on modifying the options. If your app requires runtime compilation behavior, add a custom `ICompilationService` to the service collection (overriding Mvc's default implementation). You could derive from (`DefaultRoslynCompilationService`)[https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs] and override `GetCompilationReferences` to specify your assembly list. However note that the type is in the `.Internal` namespace and we do not have guarantee compatibility between releases. Feel free to file a bug to move this to the public namespace.

**Comment by DavidEmanuelsen** [comment]
Tried the suggested approach, and as far as I can tell it gives similar behavior as when I was using the option. Debugging this shows one call to my overridden `GetCompilationReferences` as the first request comes in (before the external assembly is needed / known), and the result of this first invoke seems to be used when compiling my views using the external classes. Returning a view using this model results in a resolve-failure.

At least when I was using `Compilation.AddReferences` in the `CompilationCallback`-approach I was able to have the views resolve the types successfully in this scenario (except for the issue this bug was originally for :)).

**Comment by dougbu** [comment]
@DavidEmanuelsen you'll need to do more than override `GetCompilationReferences()`. Unfortunately you'll need to copy most of `DefaultRoslynCompilationService` and replacing the `CompilationReferences` property with something that is not lazy-inited.

**Comment by DavidEmanuelsen** [comment]
Ok, that works as expected :+1: Although I guess there will be some increased risk of breaking changes in the future as I depend on .Internal.. (ExpressionRewriter, IRazorViewEngineFileProviderAccessor). Is there any chance at all that these can be moved?

Also, will there be any significant performance hit when omitting the lazy-init?

**Comment by pranavkm** [comment]
@DavidEmanuelsen, like I mentioned earlier, feel free to file a bug to move the type (and it's dependencies) into a public namespace.

> Also, will there be any significant performance hit when omitting the lazy-init?

I think creating `MetadataReference`s reads the entire assembly into memory (along with creating Roslyn object graphs for everything within the assembly). This can be pretty memory expensive, part of the reason we lazy initialized and cached the results.
